doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -259,21 +259,41 @@ including the type hierarchy and the use of abstract DTypes.
 Class getter
 ==============================================================================
 
-To create a dtype instance from a scalar type users now call ``np.dtype`` (for
-instance, ``np.dtype(np.int64)``).
-
-To get the DType of a scalar type, we propose this getter syntax::
+To create a DType instance from a scalar type users now call
+``np.dtype`` (for instance, ``np.dtype(np.int64)``). Sometimes it is
+also necessary to access the underlying DType class; this comes up in
+particular with type hinting because the "type" of a DType instance is
+the DType class. Taking inspiration from type hinting, we propose the
+following getter syntax::
 
     np.dtype[np.int64]
 
-The notation works equally well with built-in and user-defined DTypes
-and is inspired by and potentially useful for type hinting.
+to get the DType class corresponding to a scalar type. (Note that with
+the acceptance of PEP 585 [2]_ there is precedent for
+``np.dtype[np.int64]`` referring to a *subclass* of ``np.dtype``.) The
+notation works equally well with built-in and user-defined DTypes.
 
 This getter eliminates the need to create an explicit name for every
-DType, crowding the ``np`` namespace; the getter itself signifies the type.
+DType, crowding the ``np`` namespace; the getter itself signifies the
+type. It also opens up the possibility of making ``np.ndarray``
+generic over DType class via annotations like::
+
+    np.ndarray[np.dtype[np.float64]]
+
+The above is fairly verbose, so it is possible that we will include
+aliases like::
+
+    Float64 = np.dtype[np.float64]
 
-Since getter calls won't be needed often, this is unlikely to be burdensome.
-Classes can also offer concise alternatives.
+in ``numpy.typing``, thus keeping annotations concise but still
+avoiding crowding the ``np`` namespace as discussed above. For a user
+defined DType::
+
+    class UserDtype(dtype): ...
```

Comment:
How about

```suggestion
    class UserScalar: ...
    class UserDtype(np.dtype[UserScalar]): ...
```

or

```suggestion
    T = typing.TypeVar("T")
    class UserDtype(np.dtype[T]): ...
```

Reply:
I'm not sure-in the NEP there is this line:

> The notation works equally well with built-in and user-defined DTypes

but I'm not actually sure _how_ the `__class_getitem__` would work with a user-defined scalar type. I have more been expecting that people would define new DType classes directly (maybe not even defining a corresponding scalar type!), which is why I emphasized the DType class here and left the scalars out.

Reply:
@person142 it will work for DTypes requesting it (not necessarily for all DTypes). And in that case it will work by NumPy having global `known_dtype` mapping basically.  The reason we need something in any case, is to support:
```
np.array([user_scalar])
```
without explicitly passing a `dtype=`, which I would like to support.

Reply:
Ah I see. Maybe I should include both examples then.

Reply:
Update here https://github.com/numpy/numpy/pull/17447/commits/c564fcb12789b4bc43cc86cc858be0e886824051#diff-89eb8c11ba758f8d3a296f44372c3254R293 for user-defined scalar types. AFAICT supporting `np.dtype[UserScalar]` mapping to `UserDtype` will require adding a single overload inside the NumPy stubs, which seems ok since we'll already have the `known_dtype` mapping. LMK if you see a way around that though.

Reply:
> AFAICT supporting `np.dtype[UserScalar]` mapping to `UserDtype` will require adding a single overload inside the NumPy stubs, which seems ok since we'll already have the `known_dtype` mapping.

Sounds good to me, though lets wait with this a bit until NEP 42 is implemented.

Reply:
> it will work for DTypes requesting it (not necessarily for all DTypes). And in that case it will work by NumPy having global known_dtype mapping basically. The reason we need something in any case, is to support:
> ``` python
> np.array([user_scalar])
> ```
> without explicitly passing a dtype=, which I would like to support.

Just a word of warning: I doubt we'll be able to describe such dynamic behavior with static typing.
Depending on how the overloads will be structured the dtype will probably inferred as either 
`dtype[np.object_]` or `dtype[typing.Any]`.

It's different story if the dtype is explicitly specified  (`np.array([user_scalar], dtype=UserDType`), 
then things will be quite simple.

Reply:
@BvB93, I believe you that describing it will be difficult and I do hope that most  user-dtypes will rely more on explicitly passing it. It would be an interesting question whether we can get away with not allowing auto-discovery, but the fact is that currently user-dtypes pretty much always explicitly subclass `np.generic` and honestly the only reason why they have to (aside from copy pasting the `rational` code), is exactly this auto-discovery during coercion.

Just read through these changes and they look good to me, thanks everyone! I am tempted to merge soon, or is there anything left to discuss?

Reply:
> I am tempted to merge soon, or is there anything left to discuss?

We're good on my end.

> It's different story if the dtype is explicitly specified

We can also do something in the case of

```python
x: ArrayLike[UserDtype] = [user_scalar]
np.array(x)  # Can get ndarray[UserDtype] now
```

which hopefully gives users some reasonable options.

Reply:
> It would be an interesting question whether we can get away with not allowing auto-discovery, but the fact is that currently user-dtypes pretty much always explicitly subclass `np.generic` and honestly the only reason why they have to (aside from copy pasting the `rational` code), is exactly this auto-discovery during coercion.

True, _if_ a user-dtype is based on a `np.generic` generic subclass then the type checks will work right out of the box. Thing will only become tricky (from a static typing perspective at least) when this doesn't happen.

Come to think of it, reflecting this in the example would probably be a good idea (_i.e._ make `UserScalar` a `np.generic` subclass).

Reply:
Isn’t it already a subclass in the example? (I meant for it to be.)

Reply:
I was specifically talking about the suggestion in https://github.com/numpy/numpy/pull/17447#discussion_r499269308.

Reply:
It does say "generic" currently.  Further down, I say "Scalars will be expected to derive from a NumPy scalar", but honestly, I am not sure we need to stick with it. I do not see a huge advantage, because I do not believe in "inheriting" the `dtype` attributes.  When you inherit it, you mix up contra and covariance for a container and AFAIK that just doesn't work.

The reason for saying "from a NumPy scalar", is because I do not like `np.generic`. It comes with expectations on memory layout on the NumPy side and quite frankly I don't want to use it. When we reach this stage, I would rather consider creating a `np.generic` baseclass without all of those expectations.

Does subclassing even help with typing? It seems like there is a solution that will always work and it also seems to me like users will always have to define this once for runtime and once for static typing.

Reply:
> Or from a different angle: Does subclassing really matter, as opposed to what is pretty much equivalent to an ABC you can register with?

For static typing the only important thing is that there is _some_ common baseclass or protocol for automatically annotating the likes of `np.array([user_scalar])`. 
Exactly _how_ this is accomplished doesn't really matter (well, like you said, it might matter a lot during runtime...):
* Directly subclassing `np.generic`.
* Adding `np.generic` as a superclass only during type checking.
* Adding a new ABC (which may or may not be available at runtime).
* Adding a new protocol (which, again, may or may not be available at runtime).
* _etc..._

> When we reach this stage, I would rather consider creating a `np.generic` baseclass without all of those expectations.

+1 to this suggestion!